### ISSUE TITLE
The "data" case.

### ISSUE THREAD

**Issue by TetoTheSquirrelFox** [animate1978/MB-Lab#283]
So  short story : I've coded a bunch of tools related to the shape of a model, in order to change/edit things like morphs. I also did many tools to create a new models, and everything needed for the creator to be compatible with the engine.

But, there is a big problem that I never sold, too busy with the tools to create a model from scratch (you don't need the engine for this task) : to avoid cnfusion and problems with every project done by different creators, I decided and proposed to have a special folder for the project : 
- Actually, all data are under /data/
- I proposed to have /toon/ or /my_project/ for new stuff created (here the projects would be toon and my_project).
Great, but I faced a problem that can't be solved easily, because it's sooo hardcoded in the engine that I'm not keen of changing this : everywhere in the code the engine uses file_ops.get_data_path() pretty much for everything.
So I was stuck. I don't want to touch it to avoid bugs, but because of that, I can't tell to the engine where pick-up the library.blend or the config.json for example. I did a trick for config.json, but for the blend it's impossible to change easily. I tried to find a solution and the only one I found is explained below. But I want to be sure that is OK before coding.

The solution :
------------
This solution avoids many problem like to heavily change the actual engine, that could be source of future problems.
- When you start a session you can choose a model, then you choose a template, then you create you character, before changing its shape/skin/... with dedicated tools.
![image](https://user-images.githubusercontent.com/59537161/87248433-b1107280-c459-11ea-8781-494897e74917.png)
- I propose to add a 2nd drop down list above this, that allows you to choose the "project" you want to work on. This list would be created automatically (the functions for that already exist), and basically the named showed would be the name of the project or the author if he has a big ego. We would have "MB-Lab base", "Toon", and so on.
- I just touch, in file_ops, to 1 function : get_data_path(). This function, instead of always using "data" uses instead the value of the new drop-down list. I'm not sure yet, but I think it's easy to do, as you create the character (and init the engine) **after** the choice of the template, so I think it's OK.
- This means 2 things :
1. at the end, every blend under /data/ or /toon/ must be renamed humanoid_library.blend
2. For config files, you can have more than one by project (useful when a creator wants to use a model made by someone else), it just has to be named xxx_config.json, xxx being the name of the project. The code is already changed to accept every config file put in the folder.

This way the problem is solved :
- You can have all projects you want, as you choose it since the very beginning (and when the character is created, you don't change the model anyway).
- Each project is clearly separated, so after the initialization of the character, you can't use files from other projects, and it's good.
- The only con I see for now is when a creator wants to add morphs (for example) for another project :
1. He makes a package with his morphs (like **f_af01_morphs_pseudo.json** ) for MB-Lab models. He doesn't have to create a special config.json for that, as the morphs files are loaded automatically. But : 
2. The user who wants to use it will have to add this to his actual /data/. So if he adds other and other files from another creators, it may becomes a mess after a while. But don't forget that the author will add his pseudo in the name of the files, as he can do for files that can be shared (like morphs).

So if it's OK I can try something next week.
3. That could be

**Comment by animate1978** [comment]
Yes I believe that would be ok.

I think it would help with the project, keeping things separate?

**Comment by TetoTheSquirrelFox** [comment]
Yes, I do think so because :
- You are very dependent of your base model that you're using (as an user). So, be sure that you whatever_morphs.json is for /toon/ and not for /data/ could help you to manage your data.
- I do think that this addon may be used by many many people, users and creators. Having everything separated is a good start to help to manage the database.
- A model is attached to everything that

**Comment by TetoTheSquirrelFox** [comment]
Teaser :
![image](https://user-images.githubusercontent.com/59537161/87287627-9e9a4580-c4fa-11ea-836d-d2cb5018f5bc.png)

I named this "Project" but if you want something else, no problem.

**Comment by animate1978** [comment]
Project seems to work just fine really.

Looks good.

**Comment by TetoTheSquirrelFox** [comment]
Yes, and I just finished it, so here is the related documentation.
@animate1978 : This is general documentation, the different steps for an user to sculpt the model, before finalization (as after finalization, stuff like skin, shaders, hair and so on have nothing related to the shape). I don't know where you can put it, but I think that is an important point to understand as an user and a creator, to know the mechanisms of the engine.

I did not tested extensively this. But what I did is to make a duplicate of all /data/ folder, renamed it /test/ and also renamed all models like m_ca01 as m_ca99 to be able to see differences while using the addon.

Let's go.

The steps to create a character
-----------------------
The user need to go through different steps in order to create the character he wants. They're needed to choose gender, ethnic group, age, mass, body measures and so on. Sometimes, when you can't go back. All these steps are done before finalization, an important step where user freezes the body shape of the character. AFter this point, the only things the user can do are changing shaders, adding hair, add clothes...

Step 1
---------
The user has to choose the project where he wants to work. Each project has its own models to work with, shaders, morphs and so on. The meshes are dedicated to their project, as the rest of the stuff. For example, the models available in standard with this addon are 2 human models, a male and a female, and 2 anime (Japanese style) models, a male and a female.

So, when an user launches the addon, he has this :
![image](https://user-images.githubusercontent.com/59537161/87317976-ff8c4280-c527-11ea-9490-ba5627558af1.png)

The only elements useful here are the drop-down lists "Project" and "Select"
So the first is the Project. Here we have 4 projects available : 
![image](https://user-images.githubusercontent.com/59537161/87318199-5003a000-c528-11ea-9440-9a1c68f888fe.png)
MB_Lab is the project by default, available with the addon. The others, as you can see are just examples.

We choose the default one, MB-Lab. Now we select the general type of the character.

Step 2
-------------
Now user chooses the general ethnic group of the character, the group could be Caucasian or Elf for example. Here are the groups available by default :
![image](https://user-images.githubusercontent.com/59537161/87319134-855cbd80-c529-11ea-8806-ee49da4a09e1.png)

The titles are self-explanatory. The user picks one then press "create character" button, then it's step 3.
![image](https://user-images.githubusercontent.com/59537161/87319334-c8b72c00-c529-11ea-8885-db42275ae901.png)

Step 3
--------------
When it's done, the user can't change the project, the ethnic group or the gender. He just can change its shape.
Another GUI is shown, and all operations before finalizing the character are done here.
![image](https://user-images.githubusercontent.com/59537161/87319678-39f6df00-c52a-11ea-99c2-aaea528d5bce.png)

Many tools have their dedicated documentation, we just show here the main tools to change the shape.
1. You have 3 categories that allow to change the general shape in one operation : the age, the mass and the tone.
![image](https://user-images.githubusercontent.com/59537161/87319980-9eb23980-c52a-11ea-93a7-07a4d7ff0782.png)
2. Under this you have tools that setup the character.
![image](https://user-images.githubusercontent.com/59537161/87320103-cacdba80-c52a-11ea-96cb-f4eed33e295e.png)
We will explain the "character library" tool, as it is the main tool to shape the character (see step 4).
3. Then below you have 3 tools about character design : 
![image](https://user-images.githubusercontent.com/59537161/87320367-25671680-c52b-11ea-9863-43e5924713db.png)
In step 5 we will explain the "Body measures" as it's the 2nd main tool that shapes the character.

For detailed explanation about these tools and the other, see dedicated documentation.

**Comment by animate1978** [comment]
Excellent!

I will be adding this to the docs soon as well. Just let me know when you have pushed your work to your repository and I will merge it in.

**Comment by TetoTheSquirrelFox** [comment]
No problem. As you can see, I added other stuff here. (wanted to save after a while). When I put the code in repository, that means that the documentation is done.

**Comment by TetoTheSquirrelFox** [comment]
I was thinking : this documentation is really centered to shape, and does not talk about shaders, hair, skeleton, ... that much. Maybe another step should be written about other features.

**Comment by TetoTheSquirrelFox** [comment]
@animate1978 : I discovered, while writing my post about skeleton, that this solution of adding separated folders brings bugs with creation tools (when you create/load config file and models). I'm on it and push a commit when it's done. I have to finish tomorrow, after that I'm unavailable for a while. ^^
